app/scripts/lib/rpc-method-middleware/handlers/wallet-requestPermissions.ts

```diff
@@ -0,0 +1,181 @@
+
+export const requestPermissionsHandler = {
+  methodNames: [MethodNames.RequestPermissions],
+  implementation: requestPermissionsImplementation,
+  hookNames: {
+    getAccounts: true,
+    requestPermissionsForOrigin: true,
+    requestCaip25PermissionForOrigin: true,
+  },
+};
+
+type AbstractPermissionController = PermissionController<
+  PermissionSpecificationConstraint,
+  CaveatSpecificationConstraint
+>;
+
+type GrantedPermissions = Awaited<
+  ReturnType<AbstractPermissionController['requestPermissions']>
+>[0];
+
+/**
+ * Request Permissions implementation to be used in JsonRpcEngine middleware.
+ *
+ * @param req - The JsonRpcEngine request
+ * @param res - The JsonRpcEngine result object
+ * @param _next - JsonRpcEngine next() callback - unused
+ * @param end - JsonRpcEngine end() callback
+ * @param options - Method hooks passed to the method implementation
+ * @param options.getAccounts - A hook that returns the permitted eth accounts for the origin sorted by lastSelected.
+ * @param options.requestCaip25PermissionForOrigin - A hook that requests the CAIP-25 permission for the origin.
+ * @param options.requestPermissionsForOrigin - A hook that requests permissions for the origin.
+ * @returns A promise that resolves to nothing
+ */
+async function requestPermissionsImplementation(
+  req: JsonRpcRequest<[RequestedPermissions]> & { origin: string },
+  res: PendingJsonRpcResponse<Json>,
+  _next: AsyncJsonRpcEngineNextCallback,
+  end: JsonRpcEngineEndCallback,
+  {
+    getAccounts,
+    requestPermissionsForOrigin,
+    requestCaip25PermissionForOrigin,
+  }: {
+    getAccounts: () => string[];
+    requestPermissionsForOrigin: (
+      requestedPermissions: RequestedPermissions,
+    ) => Promise<[GrantedPermissions]>;
+    requestCaip25PermissionForOrigin: (
+      requestedPermissions?: RequestedPermissions,
+    ) => Promise<
+      ValidPermission<
+        typeof Caip25EndowmentPermissionName,
+        Caveat<typeof Caip25CaveatType, Caip25CaveatValue>
+      >
+    >;
+  },
+) {
+  const { params } = req;
+
+  if (!Array.isArray(params) || !isPlainObject(params[0])) {
+    return end(invalidParams({ data: { request: req } }));
+  }
+
+  const [requestedPermissions] = params;
+  const caip25EquivalentPermissions: Partial<
+    Pick<RequestedPermissions, 'eth_accounts' | 'endowment:permitted-chains'>
+  > = pick(requestedPermissions, [
+    RestrictedMethods.eth_accounts,
+    PermissionNames.permittedChains,
+  ]);
+  delete requestedPermissions[RestrictedMethods.eth_accounts];
+  delete requestedPermissions[PermissionNames.permittedChains];
+
+  const hasCaip25EquivalentPermissions =
+    Object.keys(caip25EquivalentPermissions).length > 0;
+  const hasOtherRequestedPermissions =
+    Object.keys(requestedPermissions).length > 0;
+
+  let grantedPermissions: GrantedPermissions = {};
+  let didGrantOtherPermissions;
+
+  if (hasOtherRequestedPermissions || !hasCaip25EquivalentPermissions) {
```

Comment:
The logic here is hard to follow and diverges from the existing behaviour of `wallet_requestPermissions`.

With input such as:
```
ethereum.request({ method: 'wallet_requestPermissions', params: [{
        "eth_accounts": {},
        "wallet_snap": {
          caveats: [
            {
              type: "snapIds",
              value: { "npm:filsnap": {} },
            },
          ],
        },
      }]
});
```
The permission request is now split into two operations that can be independently accepted/rejected which is a regression in behavior. For instance, rejecting the first permission still gives you another prompt for the second permission.

Reply:
So how did this behave before this change 🤔 ? All the permissions were on a single confirmation?

Reply:
It is a regression in behavior at the UI/UX level, but not the API level.

The problem here is that `eth_accounts` and `endowment:permitted-chains` don't actually exist in the permission specifications anymore, and so they have to go through the ApprovalController directly instead for the time being.

> The permissions system does allow for partial approvals by returning back the permissions that were actually granted

Right. This behavior is preserved in this implementation. The only regression is that two approvals are shown to the user rather than one. To the dapp, nothing has changed

Reply:
They are supposed to be yeah, but that seems somewhat broken too. On main the snippet above doesn't show the Snap permission, only the install screen.

Reply:
Solved here https://github.com/MetaMask/metamask-extension/pull/29613